### PR TITLE
Add option to specify username when deploying via ssh

### PR DESCRIPTION
**Pull request by chrisbrahms**
This deals with the case of a username being specified in the the repository url in `deploydocs`. That is, if one calls `deploydocs` like this:
```julia
deploydocs(
    repo = "user@host:/path/to/repo",
)
```
deploying now works, rather than trying to push to `git@user@host:/path/to/repo`.

This came up when hosting documentation on a webserver where the `git` user already exists but I can't get access to it.

### CODE REVIEW

File: src/Documenter.jl [hunk 1]
```diff
@@ -474,7 +474,7 @@ The documentation are placed in the folder specified by `subfolder`.
 function git_push(
         root, temp, repo;
         branch="gh-pages", dirname="", target="site", sha="", devurl="dev",
-        versions, forcepush=false, deploy_config, subfolder,
+        versions, forcepush=false, deploy_config, subfolder, username="git"
```
Comment (mortenpi) [inline src/Documenter.jl:477]: `username` is not used anywhere?

File: src/Documenter.jl [hunk 2]
```diff
@@ -555,11 +555,23 @@ function git_push(
...
+            # The upstream URL to which we push new content and the ssh decryption commands.
+            upstream = "$(replace(repo, "$host/" => "$host:"))"
+        else
+            # if no user given, default to "git"
+            user = "git"
+            # Extract host from repo as everything up to first ':' or '/' character
+            host = match(r"(.*?)[:\/]", repo)[1]
+            # The upstream URL to which we push new content and the ssh decryption commands.
+            upstream = "git@$(replace(repo, "$host/" => "$host:"))"
+        end
```
Comment (mortenpi) [inline src/Documenter.jl:573]: Just a suggestion: it feels like it should be possible to do this with just a single regex, e.g.:

```julia
match(r"(?:([^@]*)@)?(.*?)[:\/]", "user@foo.com/blah.git")
```

And if the first capture group is `nothing`, then use the default username.

Comment (mortenpi) [reply src/Documenter.jl:573]: Also, could we refactor this into a function (returning a `user, upstream` tuple?) and add some unit tests, just to make sure that the code does exactly what it is supposed to?

File: src/Documenter.jl (hunk 1 above)
Comment (chrisbrahms) [reply src/Documenter.jl:477]: that was left over from an earlier attempt, thanks for catching it

File: src/Documenter.jl (hunk 2 above)
Comment (chrisbrahms) [reply src/Documenter.jl:573]: Yes, that is much more elegant (my feeble regex skills failed me). Where would be a good place to put the unit tests?

**Comment by chrisbrahms** [comment]
I've factored out the repo-disassembly into `user_host_upstream` and added some tests. These are currently in `test/deployconfig.jl`--is that the correct place?

Regarding the new function `user_host_upstream`: to make `strict` happy, should I add a `@docs` entry for it or just remove the docstring?

File: src/Documenter.jl [hunk 3]
```diff
@@ -474,7 +474,7 @@ The documentation are placed in the folder specified by `subfolder`.
 function git_push(
         root, temp, repo;
         branch="gh-pages", dirname="", target="site", sha="", devurl="dev",
-        versions, forcepush=false, deploy_config, subfolder,
+        versions, forcepush=false, deploy_config, subfolder
```
Comment (mortenpi) [inline src/Documenter.jl:477]: We generally keep the trailing commas:

```suggestion
        versions, forcepush=false, deploy_config, subfolder,
```

File: test/deployconfig.jl [hunk 4]
```diff
@@ -228,3 +228,19 @@ end end
         @test cfg === nothing
     end
 end
+##
+@testset "Remote repository paths" begin
+    uhu = Documenter.user_host_upstream("github.com/JuliaDocs/Documenter.jl.git")
+    @test uhu == ("git", "github.com", "git@github.com:JuliaDocs/Documenter.jl.git")
+
```
Comment (mortenpi) [inline test/deployconfig.jl:235]:
```suggestion

    uhu = Documenter.user_host_upstream("github.com:JuliaDocs/Documenter.jl.git")
    @test uhu == ("git", "github.com", "git@github.com:JuliaDocs/Documenter.jl.git")

    uhu = Documenter.user_host_upstream("gitlab.com/JuliaDocs/Documenter.jl")
    @test uhu == ("git", "gitlab.com", "git@gitlab.com:JuliaDocs/Documenter.jl")

```

File: src/Documenter.jl (hunk 2 above)
Comment (mortenpi) [reply src/Documenter.jl]: We can add it here: https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/src/lib/internals/documenter.md

File: src/Documenter.jl [hunk 5]
```diff
@@ -624,6 +621,29 @@ function rm_and_add_symlink(target, link)
...
+    #= the regex has three parts:
+    (?:([^@]*)@)?  matches any number of characters up to the first "@", if present,
+        capturing only the characters before the "@" - this captures the username
+    (?:([^\/:]*)[\/:]){1}  matches exactly one instance of any number of characters
+        other than "/" or ":" up to the first "/" or ":" - this captures the hostname
+    [\/]?(.*)  matches the rest of the repo, except an initial "/" if present (e.g. if
+        repo is of the form usr@host:/path/to/repo) - this captures the path on the host
+    =#
+    m = match(r"(?:([^@]*)@)?(?:([^\/:]*)[\/:]){1}[\/]?(.*)", repo)
+    isnothing(m) && error("Invalid repo path $repo")
```
Comment (mortenpi) [inline src/Documenter.jl:640]: This is failing on 1.0 (we're not depending on Compat):

```suggestion
    (m === nothing) && error("Invalid repo path $repo")
```

**Comment by mortenpi** [comment]
> Regarding the new function `user_host_upstream`: to make `strict` happy, should I add a `@docs` entry for it or just remove the docstring?

We can add it here: https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/src/lib/internals/documenter.md

**Comment by mortenpi** [comment]
Good to go as far as I can tell. Thanks for the contribution @chrisbrahms!

**Pull request merged**